File: 백준/js/boj_2798.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs")
  .readFileSync(filePath)
  .toString()
  .trim()
  .split("\n");

const [N, M] = input.shift().split(" ").map(Number);
const card = input[0].split(" ").map(Number);

const check = new Array(N).fill(false);
const cardArr = [];
let answer = 0;

function dfs(depth) {
  if (depth === 3) {
    let sum = cardArr.reduce((s, v) => s + v);
    if (sum > answer && sum <= M) {
      answer = sum;
    }
    return;
  }
  for (let i = 0; i < N; i++) {
    if (!check[i]) {
      check[i] = true;
      cardArr.push(card[i]);
      dfs(depth + 1);
      check[i] = false;
      cardArr.pop();
    }
  }
}

dfs(0);
console.log(answer);

File: 백준/python/boj_2798.py
import sys

sys.stdin = open("input.txt", "r")

N, M = map(int, input().split(" "))
card = list(map(int, input().split()))

# 백트래킹 풀이
answer = 0
cardArr = []

def dfs(idx, cnt):
    global answer
    if cnt == 3:
        total = sum(cardArr)
        if total > answer and total <= M:
            answer = total
        return

    for i in range(idx, N):
            cardArr.append(card[i])
            dfs(i+1, cnt+1)
            cardArr.pop()

dfs(0,0)
print(answer)

# 조합 풀이
from itertools import combinations
answer = 0

for cards in combinations(card, 3):
    total = sum(cards)
    if total > answer and total <= M:
        answer = total

print(answer)

